From 5b5bfadd3ba5ae10e1b4fb1cb5b80bf101204c12 Mon Sep 17 00:00:00 2001
From: "yadi.hu@windriver.com" <yadi.hu@windriver.com>
Date: Wed, 25 May 2016 10:35:51 +0800
Subject: [PATCH] rtc:fix system pm_power_off on rtc driver

After issuing "poweroff" or "shutdown -h 0" commands
the board is reports the following error message:

=====================================
[ BUG: lock held at task exit time! ]
3.10.62-ltsi-WR6.0.0.17_standard #1 Tainted: G W
-------------------------------------
halt/634 is exiting with locks still held!
1 lock held by halt/634:
 #0: (reboot_mutex){......}, at: SyS_reboot+0xcc/0x1d0

Before calling do_exit(0), there is a reboot_mutex lock that is taken
in reboot syscall. do_exit() function is , checking for any locks held
by the processor and if yes then it is printing a BUG_ON from
print_held_locks_bug() function along with the locks held information.
Even though the BUG triggers, the system is going to power off because
The hardware has been triggered and We will have already stopped all
other CPUs at this point.

Signed-off-by: Hu <yadi.hu@windriver.com>
---
 drivers/rtc/rtc-snvs.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/drivers/rtc/rtc-snvs.c b/drivers/rtc/rtc-snvs.c
index 6c6b684..215714d 100644
--- a/drivers/rtc/rtc-snvs.c
+++ b/drivers/rtc/rtc-snvs.c
@@ -250,6 +250,9 @@ static void snvs_poweroff(void)
 	value = readl(snvs_base + SNVS_LPCR);
 	/* set TOP and DP_EN bit */
 	writel(value | 0x60, snvs_base + SNVS_LPCR);
+
+	local_irq_disable();
+	while (1);
 }
 
 static int snvs_rtc_probe(struct platform_device *pdev)
-- 
1.7.5.4

